### ISSUE TITLE
Road to v1

### ISSUE THREAD

**Issue by jmeistrich** [LegendApp/legend-list#28]
I think we're feature complete for version 1 :). I could use some help with making sure all of these things are good. Please let me know if you find any more issues!

### Issues
Issues to be done for v1 are tagged as 1.0:
https://github.com/LegendApp/legend-list/issues?q=is%3Aissue%20state%3Aopen%20label%3A1.0

### Features possible before v1
- [ ] Sticky headers #72 

### Future features beyond v1
- Web implementation
- Possibly more DX improvements
- More performance testing and optimizations

### Things to check
- [ ] Are all of the API types correct?
- [ ] Are there props from ScrollView that should be omitted?
- [ ] Are there still any props/features from FlatList/FlashList that should be added?
- [ ] Are there any other bugs that there's not already an Issue for?
- [ ] Do more perf testing to make sure it's actually fast in all scenarios, and optimize it more if not

**Comment by imranbarbhuiya** [comment]
I want to request a feature where `onEndReached` work both ways (reach top and bottom). Currently flashlist/flatlist/scrollview all work for bottom reach, not top

**Comment by sabaturgay** [comment]
Thank you for the great work on LegendList! It's been a pleasure using this component, which improves on FlatList in many ways. I would like to ask if there is a plan to add **_numColumns_** prop. This addition would simplify migration and enhance flexibility for grid-based UIs.

Looking forward to your thoughts. Thanks again for your hard work!

**Comment by jmeistrich** [comment]
@sabaturgay Yes, I'll add a numColumns prop soon 👍

**Comment by jmeistrich** [comment]
I just merged in a big change to add `maintainVisibleContentPosition` and a bunch of optimizations. So now the focus is on stabilizing and documenting towards a v1 release 🎉. I'll release a `0.5` with those changes tomorrow (in Germany time because I'm in Berlin right now).

Please let me know if you see any remaining issues, anywhere from serious bugs to missing features to typos.

If anyone wants to add more examples it would be much appreciated! @mbilenko-florio added a chat example in #37 which is awesome 😍. I restructured the example app to have a list of examples on the home page. So you can just PR an addition to `example/app/(tabs)/index.tsx` if you'd like.

Thank you!

**Comment by assimelha** [comment]
Amazing work @jmeistrich!

One prop that's needed especially for the react-native-bottom-sheet lib is [renderScrollComponent](https://shopify.github.io/flash-list/docs/usage/#renderscrollcomponent)

**Comment by jmeistrich** [comment]
@assimelha Oh I actually added a `renderScrollComponent` prop in the latest version but forgot to add it to the documentation. I did some quick tests and it seemed to work, but I'm not sure if it's correct to support bottom sheet. What do you think? Do you want to try it and see if it works?

**Comment by brunocentanaro** [comment]
Awesome library! Really nice work.
In my use case one prop

**Comment by assimelha** [comment]
> @assimelha Oh I actually added a `renderScrollComponent` prop in the latest version but forgot to add it to the documentation. I did some quick tests and it seemed to work, but I'm not sure if it's correct to support bottom sheet. What do you think? Do you want to try it and see if it works?

Awesome! I'll give it a try this afternoon and report back

**Comment by rawrmonstar** [comment]
For me to be able to use this over Flashlist, I need the ability to have some items span multiple columns. At the moment I implement this with `getItemType` and `overrideItemLayout` that doesn't seem possible with the current props provided by LegendList

**Comment by r0b0t3d** [comment]
https://github.com/user-attachments/assets/67505954-b09e-47c4-a66c-52e14d7323a1

Reloading in the top of the list is fine, but when I scroll down and the list got reloaded, I can see blank screen. Scroll the list again and it shows up

**Comment by 2Senn** [comment]
Flashlist had an issue with nested lists not scrolling. This is a problem with Flatlist as well unless you use the Flatlist from react-native-gesture-handler. I am not sure why this happens, but I hope LegendList will support nested scrolls like a horizontal image gallery.

**Comment by michbil** [comment]
> legend-list.mp4 
> Reloading in the top of the list is fine, but when I scroll down and the list got reloaded, I can see blank screen. Scroll the list again and it shows up

Thanks for you bug report, the issue looks similar to https://github.com/LegendApp/legend-list/issues/77 pls follow the thread for solution.

**Comment by jmeistrich** [comment]
> Flashlist had an issue with nested lists not scrolling. This is a problem with Flatlist as well unless you use the Flatlist from react-native-gesture-handler. I am not sure why this happens, but I hope LegendList will support nested scrolls like a horizontal image gallery.
> 
> Great work ! <3

It does already. See the Movies example in the example app in the repo 😃

**Comment by diegolmello** [comment]
Looks like a great project!
The blockers for us are the exame items listed as issues:
- https://github.com/LegendApp/legend-list/issues/83
- https://github.com/LegendApp/legend-list/issues/113
- https://github.com/LegendApp/legend-list/issues/117

What's the behavior when the list is inverted and the users is on VoiceOver?
FlatList/FlashList inverts the list, so the messages are inverted on the screen reader, so the UX is awkward.
Thanks!

**Comment by Teczer** [comment]
In the context of a chat application, the data for chat messages is currently ordered incorrectly.

The messages are expected to be paginated from the most recent to the oldest, meaning the newest messages should appear first in the list. As users scroll up, they should encounter older messages.

However, the current implementation has the messages ordered from oldest to newest, which is the reverse of the expected order.

This needs to be corrected so that the most recent messages are displayed first, aligning with the typical user experience in chat applications where users see the latest messages at the top and can scroll up to view previous ones.

**Comment by diegolmello** [comment]
@Teczer Can't you invert your messages array before adding to the list? Does that make sense for your use case?

**Comment by jmeistrich** [comment]
@Teczer I disagree. Content in apps is always ordered from top to bottom, usually in ascending order. Chat UIs being the one place where everything renders backwards is a bug. Rendering them in ascending order is correct. You add a new message, it goes to the end of the list at the bottom. Adding a message by splicing it into the top of the list does not make sense.

You're used to other lists that use `inverted`, which makes it the opposite of normal, to render message UIs. But inverted causes all sorts of weird behavior and bugs, like with screen readers as @diegolmello mentioned above. So we purposefully do not have an `inverted` prop, and have `alignItemsAtEnd` and `maintainScrollAtEnd` instead, for making chat UIs.

If messages are coming in from your server in the opposite order, that's fine, just reverse the array before rendering them.

**Comment by jmeistrich** [comment]
And @diegolmello sorry I didn't see your message before! As of today in beta.28 all of those blockers should now be fixed. And as I just mentioned, there's no inverted, so screen reader should be fine :)

**Comment by diegolmello** [comment]
Awesome, @jmeistrich. Going to raise the priority of this migration! 🙏🏼
